content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,152 @@
+
+But, what if I tell you there are helpful approaches to write null-safe code. In particular, if you are using the Spring Framework. Wouldn't that be great news?
+
+Luckily, Spring Framework offers some annotations to solve this exact problem. 
+In this article, we will learn how to use these annotations to write null-safe code using [Spring Boot](https://reflectoring.io/categories/spring-boot/).
+
+## Null-Safety Annotations in Spring
+
+Under the `org.springframework.lang` Spring core package, there are 4 such annotations, `@NonNull`, `@NonNullFields`, `@Nullable`, and `@NonNullApi`. These are [meta-annotations](https://dzone.com/articles/what-are-meta-annotations-in-java) that use the famous (even though dormant at present) [JSR-305](https://jcp.org/en/jsr/detail?id=305) annotations underneath. 
+
+**Popular IDEs like [Eclipse](https://www.eclipse.org/ide/) and [IntelliJ IDEA](https://www.jetbrains.com/idea/) can understand these annotations.** It can warn the developers of potential issues during compile time. 
+
+We are going to use IntelliJ IDEA in this tutorial. Let us find out more with some code examples.
+
+To create the base project, we have taken help from the [Spring Initializr](https://start.spring.io/). The Spring Boot starter is all we need, no need to add any extra dependencies.
+
+## `@NonNull`
+
+Let's use a plain `Employee` class to explain the annotations:
+
+```java
+package io.reflectoring.nullsafety;
+
+// imports 
+
+class Employee {
+  String id;
+  String name;
+  LocalDate joiningDate;
+  String pastEmployment;
+
+  // standard constructor, getters, setters
+}
+```
+
+Mostly the `id` field is going to be a non-nullable value. So, to avoid any potential `NullPointerException` we can mark this field as [`NonNull`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNull.html):
+
+
+```java
+class Employee {
+
+  @NonNull 
+  String id;
+
+  //...
+}
+```
+Now, if we accidentally try to set the value of `id` as null anywhere in the code, the IDE will show a compilation warning:
+
+{{% image alt="IDE warning for NonNull" src="images/posts/spring-boot-null-safety-annotations/nonnull-ide-warning.png" %}}
+
+Pretty awesome, isn't it?
+
+**The `NonNull` annotation can be used at the method, parameter, or field level.** 
+
+At this point, you might be thinking, what if a class has more than one non-null field. Would it not be too wordy if we have to add a `@NonNull` annotation before each of these? 
+
+We can solve this problem by using the [`@NonNullFields`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullFields.html) annotation. Let us see how in the next section.
+
+## `@NonNullFields`
+
+Let us create a `package-info.java` file to apply the non-null field checks at the package level. This file will contain the root package name with `@NonNullFields` annotation:
```

Comment:
This makes it seem that `@NonNullFields` is only applicable to packages. That's not true, is it? Please add a table like mentioned in the comment above.

Reply:
`@NonNullFields` is meant for packages only. 
[Source](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/NonNullFields.html)